KEYCLOAK-8996: Provide a way to set a responder certificate in OCSP/X509 Authenticator

rmartinc opened the pull request:
Fix for KEYCLOAK-8996.

The PR adds a new field in the X509 authenticator configuration to set a responder cert using the PEM format (similar to what is done in the SAML broker, the code for reading the cert is re-used from that part). There are two things to take into account:

* The responder cert is only used if responder URI is set (not when using revocation endpoints, the cert is not passed in this check). I preferred to it this way (but this can be changed).
* A new `Text` type has been added to show a text-area in the X509 configuration. The typical `String` field was horrible.

rmartinc committed:
KEYCLOAK-8996: Provide a way to set a responder certificate in OCSP/X509 Authenticator

stianst commented:
Thanks for your PR. We are not able to review and merge this right now as we are not accepting new features and enhancements for 4.x and we will not be able to merge this until we open up for 5.x development early next year.

rmartinc commented:
@stianst OK Stian, no problem. We'll manage this later to 5.x, when the 4.x release is over.

rmartinc commented:
As this is on hold maybe we can add another improvement. I'm a bit concerned about how CRLs are managed. It seems that the CRL file/url is loaded in every check. Maybe caching the CRL somehow would be better. I was thinking about just reloading it if the configuration changes (there are more complicated ways of doing the same). This way the admin has a way to force a reload of the CRL (just doing a kcadm.sh or a curl to modify the X509 configuration after modifying the CRL file itself).

Is it possible to see if the configuration has changed in the authenticator? For example doing something like:

```java
context.getAuthenticatorConfig() != cachedConf
```
Or it is not possible to see if there is a new configuration.

s7726 commented:
Seems like a better way to approach this would be to just allow the existing configuration options.
https://docs.oracle.com/javase/6/docs/technotes/guides/security/certpath/CertPathProgGuide.html#AppC

Somewhere around here:
https://github.com/keycloak/keycloak/blob/master/common/src/main/java/org/keycloak/common/util/OCSPUtils.java#L392-L395

rmartinc commented:
Hi Gavin,

Keycloak is not using the default Java OCSP implementation, so that options don't apply here. It is using the BouncyCastle implementation. And in BC you have to pass the certificate to the method [here](https://github.com/keycloak/keycloak/blob/a2084b64b7418efa748788248e061f42a8d5d709/services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java#L191). I have modified the provider configuration to let configure a different responderCert per instance.

mposolda commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/AbstractX509ClientCertificateAuthenticatorFactory.java:
```diff
@@ -155,6 +157,12 @@
+        ocspResponderCert.setType(TEXT_TYPE);
+        ocspResponderCert.setName(OCSPRESPONDER_CERTIFICATE);
+        ocspResponderCert.setLabel("OCSP Responder Certificate");
+        ocspResponderCert.setHelpText("Certificate used by the responder to sign the responses. The certificate should be in PEM format without BEGIN and END tags. Only used if OCSP Responder URI is set.");
```
Is it possible to specify more clearly that this is not mandatory field? For example add another sentence to help text with something like "If this is not set, then the certificate of the CA, who issued the client certificates is used to verify the signature of OCSP response."

mposolda commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java:
```diff
@@ -675,6 +682,21 @@ public GotCRLRelativePath cRLLoader(CRLLoaderImpl cRLLoader) {
+                public GotOCSP oCSPResponseCertificate(String responderCert) {
+                    if (responderCert != null && !responderCert.isEmpty()) {
+                        try {
+                            _responderCert = XMLSignatureUtil.getX509CertificateFromKeyInfoString(responderCert);
```
I wonder what is the performance impact of this? I think it will be good to have a possibility to "cache" the X509Certificate object, so it doesn't need to be parsed from PEM on every request. However this is probably not for this PR, but rather an improvement for the future. Just mentioning it in this PR when thinking about it :)

mposolda reviewed (commented):
@rmartinc Thanks, looks quite good to me. And sorry it took longer to review. Just added one minor comment to improve the helpText, but besides that, it is fine IMO.

rmartinc commented:
Thanks @mposolda,

I have added a more detailed description of the field (I have more or less copied the one of the java property `ocsp.responderCertIssuerName` which is very detailed). Regarding the second issue I also think that some caching is necessary, but better if it is done to the whole X509 configuration (CRLs inclusive which can be more expensive). I have rebased the branch (although no conflicts existed) and tested with x509 tests again. Let me know if you need something more.

mposolda commented:
@keycloak-ci-bot test
